skcosmo/feature_selection/voronoi_fps.py

```diff
@@ -0,0 +1,163 @@
+        for n in range(n_pad):
+            self.idx_in_voronoi.append(np.array([]))
+        self.number_in_voronoi = np.pad(
+            self.number_in_voronoi, (0, n_pad), "constant", constant_values=0
+        )
+        self.sel_d2q_ = np.pad(self.sel_d2q_, (0, n_pad), "constant", constant_values=0)
+        self.furthest_point = np.concatenate(
+            (self.furthest_point, np.zeros(n_pad, dtype=int))
+        )
+
+    def _get_active(self, X, last_selected):
+        # because we haven't called super() yet, self.n_selected_ has not been
+        # incremented yet
+
+        f_active = np.zeros(self.n_features_to_select, bool)
+
+        """must compute distance of the new point to all the previous FPS. Some
+            of these might have been computed already, but bookkeeping could be
+            worse that recomputing (TODO: verify!)
+            """
+        # calculation in a single block
+        self.sel_d2q_[: self.n_selected_] = (
+            self.norms_[self.selected_idx_[: self.n_selected_]]
+            + self.norms_[last_selected]
+            - 2 * (self.X_selected_[:, : self.n_selected_].T @ X[:, last_selected])
+        ) * 0.25
+
+        for ic in range(self.n_selected_):
+            # empty voronoi, no need to consider it
+            if self.number_in_voronoi[ic] > 1:
+                r2 = max(self.haussdorf_[self.idx_in_voronoi[ic]])
+                if self.sel_d2q_[ic] < r2:
+                    # these voronoi cells need to be updated
+                    f_active[ic] = True
+
+        return f_active
+
+    def _calculate_distances(self, X, last_selected, **kwargs):
+        # n_selected has not been incremented, so index of new voronoi is
+        # n_selected
+        self.eligible_[last_selected] = False
+        self.haussdorf_[last_selected] = 0.0
+
+        if self.n_selected_ == 0:
+            self.haussdorf_ = super()._calculate_distances(X, last_selected)
+            updated_points = np.arange(X.shape[-1], dtype=int)
+
+        else:
+            f_active = self._get_active(X, last_selected)
+            if len(np.where(f_active)[0]) > 0:
+                if (
+                    np.sum(self.number_in_voronoi[f_active]) / X.shape[1]
+                    > VORONOI_CUTOFF_FRACTION
+                ):
+                    new_dist = super()._calculate_distances(X, last_selected)
+                    updated_points = np.where(new_dist < self.haussdorf_)[0]
+                    np.minimum(self.haussdorf_, new_dist, self.haussdorf_)
+                    f_active = np.full(self.n_selected_, False)
+                    f_active[self.vlocation_of_idx[updated_points]] = True
+                elif len(np.where(f_active)[0]) > 0:
+                    new_dist = self.haussdorf_.copy()
+                    active_points = np.concatenate(
+                        np.array(self.idx_in_voronoi, dtype=object)[f_active]
+                    )
+
+                    new_dist[active_points] = (
```

Comment:
I will ask the same question again as last time. Why don't we want to check that (self.haussdorf_[i]<self.sel_d2q_[i])? We already have these distances calculated, and such a check gives us a chance to reduce the number of calculations.

Reply:
As discussed, we can add another filter between lines 110 and 112 to reduce the set of active points to only those whose distance from their voronoi center is greater than the d2q of the center to the last selected point.